File: android/src/main/java/com/kakaotam/rnsdk/link/LinkApiClientModule.kt
package com.kakaotam.rnsdk.link

import com.facebook.react.bridge.*
import com.kakao.sdk.link.LinkClient
import com.kakao.sdk.link.model.ImageUploadResult
import com.kakao.sdk.link.model.LinkResult
import com.kakaotam.rnsdk.common.PromiseTransformer
import com.kakaotam.rnsdk.common.ReactToParam
import com.kakaotam.rnsdk.common.SdkInitializer
import com.kakaotam.rnsdk.link.model.*

class LinkApiClientModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "LinkApiClientModule"
    }

    @ReactMethod
    private fun customTemplate(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, CustomTemplateParam::class.java)!!
        LinkClient.instance.customTemplate(
            reactContext.currentActivity!!,
            p.templateId,
            p.templateArgs,
            p.serverCallbackArgs
        ) f@{ linkResult: LinkResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(linkResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun defaultTemplate(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, DefaultTemplateParam::class.java)!!
        LinkClient.instance.defaultTemplate(
            reactContext.currentActivity!!,
            p.templateObject!!,
            p.serverCallbackArgs
        ) f@{ linkResult: LinkResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(linkResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun scrapImage(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ScrapImageParam::class.java)!!
        LinkClient.instance.scrapImage(
            p.imageUrl,
            p.secureResource
        ) f@{ imageUploadResult: ImageUploadResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(imageUploadResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun scrapTemplate(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ScrapTemplateParam::class.java)!!
        LinkClient.instance.scrapTemplate(
            reactContext.currentActivity!!,
            p.requestUrl,
            p.templateId,
            p.templateArgs,
            p.serverCallbackArgs
        ) f@{ linkResult: LinkResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(linkResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun uploadImage(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, UploadImageParam::class.java)!!
        LinkClient.instance.uploadImage(
            p.image,
            p.secureResource
        ) f@{ imageUploadResult: ImageUploadResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(imageUploadResult))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }
}

File: ios/KakaoSdkLink/LinkApiClientModule.swift
//
//  LinkApiClientModule.swift
//  react-native-kakao-sdk-all
//
//  Created by kakao-tam on 2022/05/06.
//

import Foundation

import KakaoSDKLink

@objc(LinkApiClientModule)
class LinkApiClientModule: NSObject {
 
    @objc(customTemplate:withResolve:withReject:)
    func customTemplate(_ param:NSDictionary,
                        resolve:@escaping RCTPromiseResolveBlock,
                        reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(CustomTemplateParam.self)!
            LinkApi.shared.customLink(templateId: p.templateId,
                                      templateArgs: p.templateArgs,
                                      serverCallbackArgs: p.serverCallbackArgs) {
                (linkResult: LinkResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(linkResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(defaultTemplate:withResolve:withReject:)
    func defaultTemplate(_ param:NSDictionary,
                         resolve:@escaping RCTPromiseResolveBlock,
                         reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(DefaultTemplateParam.self)!
            LinkApi.shared.defaultLink(templateObject: p.templateObject,
                                       serverCallbackArgs: p.serverCallbackArgs) {
                (linkResult: LinkResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(linkResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(scrapImage:withResolve:withReject:)
    func scrapImage(_ param:NSDictionary,
                    resolve:@escaping RCTPromiseResolveBlock,
                    reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ScrapImageParam.self)!
            LinkApi.shared.imageScrap(imageUrl: p.imageUrl,
                                      secureResource: p.secureResource) {
                (imageUploadResult: ImageUploadResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(imageUploadResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(scrapTemplate:withResolve:withReject:)
    func scrapTemplate(_ param:NSDictionary,
                       resolve:@escaping RCTPromiseResolveBlock,
                       reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ScrapTemplateParam.self)!
            LinkApi.shared.scrapLink(requestUrl: p.requestUrl,
                                     templateId: p.templateId,
                                     templateArgs: p.templateArgs,
                                     serverCallbackArgs: p.serverCallbackArgs) {
                (linkResult: LinkResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(linkResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(uploadImage:withResolve:withReject:)
    func uploadImage(_ param:NSDictionary,
                     resolve:@escaping RCTPromiseResolveBlock,
                     reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(UploadImageParam.self)!
            LinkApi.shared.imageUpload(image: p.image,
                                       secureResource: p.secureResource) {
                (imageUploadResult: ImageUploadResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(imageUploadResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
}
